File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/001-two-sum-approach3.ts
function assert(condition: boolean, msg: string = ''): void {
    if (!condition) throw new Error(msg || 'Assertion failed');
}

function twoSum(nums: number[], target: number): number[] {
    const seen = new Map<number, number>();
    for (let i = 0; i < nums.length; i++) {
        const complement = target - nums[i];
        if (seen.has(complement)) return [seen.get(complement)!, i];
        seen.set(nums[i], i);
    }
    return [];
}

assert(JSON.stringify(twoSum([2, 7, 11, 15], 9)) === JSON.stringify([0, 1]));
assert(JSON.stringify(twoSum([3, 2, 4], 6)) === JSON.stringify([1, 2]));
assert(JSON.stringify(twoSum([3, 3], 6)) === JSON.stringify([0, 1]));
assert(JSON.stringify(twoSum([1, 2, 3, 4, 5], 9)) === JSON.stringify([3, 4]));
assert(JSON.stringify(twoSum([0, 4], 4)) === JSON.stringify([0, 1]));
console.log("all tests pass");

File: src/content/docs/topics/cs/coding-problems/arrays-and-hashing/001-two-sum-approach3.py
def two_sum(nums: list[int], target: int) -> list[int]:
    seen = {}
    for i, x in enumerate(nums):
        complement = target - x
        if complement in seen:
            return [seen[complement], i]
        seen[x] = i
    return []

assert two_sum([2, 7, 11, 15], 9) == [0, 1]
assert two_sum([3, 2, 4], 6) == [1, 2]
assert two_sum([3, 3], 6) == [0, 1]
assert two_sum([1, 2, 3, 4, 5], 9) == [3, 4]
assert two_sum([0, 4], 4) == [0, 1]
print("all tests pass")
